### ISSUE TITLE
max-statements-per-line misses braceless if-else statement

### ISSUE THREAD

**Issue by glenjamin** [eslint/eslint#5926]
**What version of ESLint are you using?**
2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```json
{
    "rules": {
        "max-statements-per-line": ["error", { "max": 1 }]
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

When running eslint on the following file:

```js
var a = 1;

if (a) a *= 1; else b *= 1;

if (a) { a *= 1; } else { b *= 1; }
```

**What did you expect to happen?**

I would have expected both of these if statements to be flagged as having two many statements.

**What actually happened? Please include the actual, raw output from ESLint.**

I get this:

```
/private/tmp/blah/temp.js
  5:10  error  This line has too many statements. Maximum allowed is 1  max-statements-per-line
```

Is this intentional, or an oversight? If the latter, I can probably look at sending in a PR.

**Comment by scriptdaemon** [comment]
The rule ignores the bodies of the `if` and the `else` if they not block statements, so that's why the first line isn't caught. The intent was to allow lone `if` statements with a single statement as the body be on the same line, however, I think there should be an exception made for the case you present so that it catches, since that is definitely more than one statement.

**Comment by kaicataldo** [comment]
Was able to replicate locally with the master branch. Seems like a bug to me.

**Comment by glenjamin** [comment]
I've got a branch ready for this, not opened a PR yet as I'm not sure on the procedure for stuff before it's `accepted`. https://github.com/glenjamin/eslint/tree/count-else-as-a-statement

The fix is a little weird, because the existing rule wants to treat `if (x) var y = 1` as a single statement, when *technically* it's two. But I think I get the general feeling behind this.

On a similar note, the current tests don't have any examples of `return`, but I was thinking it might be nice to treat a function body that only has 1 statement as an expression, so the following would be allowed:

```js
return function(x) { return x * 1; };
```

**Comment by kaicataldo** [comment]
@glenjamin Awesome, thanks for looking into this! Generally we recommend waiting until it's accepted so that we're all in agreement on what needs to change before someone spends time working on a fix. I just want to make sure we get another pair of eyes from @eslint/eslint-team on this before marking it as accepted.

**Comment by mysticatea** [comment]
Thank you for this issue. I'm sorry for late response.
I also think this is a bug. I added `accepted` label.

/cc @btmills @michaelficarra

**Comment by glenjamin** [comment]
PR is now open, is it worth me raising another issue for the `return function(x) { return x * 1; };` in https://github.com/eslint/eslint/issues/5926#issuecomment-213936189

**Comment by scriptdaemon** [comment]
Unrelated issues belong in their own issue.

To me it doesn't make sense to count that as 1 statement, but you can create the issue to get other opinions.

The reason we allow `if (condition) statement;` (and not `if (condition) { statement; }`) as a single statement is that keeping it on a single line is a very common practice (guard clauses, etc.). What you're proposing clearly breaks the rule to me, and should be counted as separate statements in my mind. I don't think a user would expect that kind of behavior by default (as opposed to my previous example). Plus, it raises the question of what other exceptions would we allow to make sure the rule is consistent.

**Comment by btmills** [comment]
To summarize and confirm my understanding, given `"max-statements-per-line": ["error", { "max": 1 }]`:

1. `var x = 1;` This is clearly fine.
1. `if (x) x *= 1;` To support single-line `if`s, we allow this right now. Everyone agrees that this case should remain valid?
1. `if (x) x *= 1; else x /= 1;` This case currently passes, but everyone thinks it should be a warning?
1. `if (x) { x *= 1; }` Currently not allowed, and that's the correct behavior.
1. `if (x) { x *= 1; } else { x /= 1; }` Currently not allowed, and that's the correct behavior.

If my understanding is correct, then I agree that (3) should fail. Though #5993 is currently labeled as `Breaking:`, I think we can consider this an oversight in the implementation and treat it as a `Fix:` instead.  The issue is labeled as a "bug" right now, which seems right to me. @nzakas or @ilyavolodin am I on the right track there?

**Comment by ilyavolodin** [comment]
I think I agree. This seems more like an oversight. `if` and `else` should cont as two statements on the same line.

**Comment by glenjamin** [comment]
Ah ok, so I should re-label as fix?

**Comment by ilyavolodin** [comment]
@glenjamin Let's wait to hear from @nzakas first, if he agrees, then it's a non-breaking change and changing commit message would allow us to merge it in now, if not - then this has to wait till 3.0 release.

**Comment by nzakas** [comment]
It seems like the consensus is that this is a bug in that the current behavior was unexpected, so I would consider this a bug and not a breakdown breaking change.
